src/controllers/video.controller.js

```diff
@@ -10,33 +10,135 @@ import {uploadOnCloudinary} from "../utils/cloudinary.js"
+                videoFile:video.url,
+                thumbnail:thumbnail.url,
+                title,
+                description,
+                duration:video.duration,
+                owner: ref.user._id
+            })
+            if(!uploadedVideo)
+                {
+                    throw new ApiError(500,"There was a problem while uploading video")
+                }
+        return res.status(200).json(new ApiResponse(200,uploadedVideo,"Video was uploaded successfully"))
 })
 
 const getVideoById = asyncHandler(async (req, res) => {
     const { videoId } = req.params
     //TODO: get video by id
+    if(!videoId)
+        {
+            throw new ApiError(400,"Video Id is required")
+        }
+    const video=await Video.findById(videoId)
+    if(!video)
+        {
+            throw new ApiError(400,"No video exists with the id")
+        }
+        return res.status(200).json(new ApiResponse(200,video,"Video returned successfully"))
+
 })
 
 const updateVideo = asyncHandler(async (req, res) => {
     const { videoId } = req.params
+    const {title,description}=req.body
     //TODO: update video details like title, description, thumbnail
+    if(!videoId)
+        {
+            throw new ApiError(400,"Video Id is required")
+        }
+    if([title,description].some(fields=>fields.trim()===""))
+        {
+            throw new ApiError(400,"Title and Descreption is required")
+        }
+
+        const thumbnailLocalPath=req.file?.path
+        if(!thumbnailLocalPath)
+            {
+                throw new ApiError(400,"Thumbnail is required")
+            }
+        const thumbnail=await uploadOnCloudinary(thumbnailLocalPath)
+
+        if(!thumbnail)
+            {
+                throw new ApiError(500,"There was Problem while uploading Thumbnail on cloud")
+            }
 
+        const updatedVideo=await Video.findByIdAndUpdate(videoId,{
+            $set:{
+                 title,description,thumbnail:thumbnail.url
+            }
+        },
+        {
+            new:true
+        }
+    )
+    if(!updatedVideo)
+        {
+            throw new ApiError(500,"Video was not Updated Due to some error")
+        }
+        
+        return res.status(200).json(new ApiResponse(200,updatedVideo,"Video was updated successfully"))
 })
 
 const deleteVideo = asyncHandler(async (req, res) => {
     const { videoId } = req.params
-    //TODO: delete video
+    //TODO: delete video\
+    if(!videoId)
+        {
+            throw new ApiError(400,"Video id is required")
+        }
+    const deletedVideo=await Video.findByIdAndDelete(videoId)
+    if(!deleteVideo)
+        {
+            throw new ApiError(500,"There was a problem while deleting the video")
+        }
+
+        return res.status(200).json(new ApiResponse(200,{},"Video was deleted Successfully"))
+
 })
 
 const togglePublishStatus = asyncHandler(async (req, res) => {
     const { videoId } = req.params
+    if(!videoId)
```

Comment:
**Use this** it will not give reference error:

` const currentVideo = await Video.findById(videoId);
  
  if(!currentVideo){
    throw new ApiError(404,"Video does not exists")
  }

  const updatedVideo = await Video.findByIdAndUpdate(
    videoId,
    {
      $set: {
        isPublished: !currentVideo.isPublished,
      },
    },
    {
      new: true,
    }
  );`

Reply:
Basically you find video by findbyid then again you call findbyidAndUpdate again that decrease your code optimization.

Basically one MongoDB hand shaking more than enough hear.

Reply:
we can achieve the same goal by leveraging the aggregation pipeline feature of the findOneAndUpdate 

const updatedVideo = await Video.findOneAndUpdate(
    { _id: videoId },
    [
      {
        $set: {
          isPublished: { $not: "$isPublished" }
        }
      }
    ],
    { new: true, runValidators: true }
  );

Reply:
👍